articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
+
+Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
+
+On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
```

Comment:
@wjwwood wrote:
> For the second point, I have to disagree. I can think of a lot of situation where error recovery is more complicated than close and reopen. I can give some anecdotes:
> [..]

@wjwwood: I think what @doudou is trying to say is that the things you described should not be considered the responsibility of the nodes themselves, but are configuration / initialisation issues that the entire (sub)system should be aware of or really in control of. By introducing (or explicitly allowing) node behaviour that is invisible from the outside, it would seem external coordination becomes much harder, and the system less transparent.

If I understand @doudou correctly, I think the behaviour you described in your two examples is exactly the kind of thing that can be modelled very well by a suitable combination of external (to the node) coordination and configuration:

 - `ax2550`: configure node for 'configuration mode', start it. Node fails (motor is not in expected mode), signals coordinator (implicitly, by transitioning to error state). Now configure node for 'command mode', start it again. Node should succeed.
 - `rmp`: start node, node fails (hw in unexpected state (not properly shutdown)). Coordinator waits a bit, then reinits and restarts node.

> Now you can argue that this is the devices fault or that I can do the delay in the destructor or something like that, but it probably be better to just have the node (and not the supervisor) do the delay in an error state before telling the supervisor to try the transition again.

If you see 'the supervisor' as something completely external to (and far removed from) 'the node', then yes: otherwise that (single?) supervisor would need knowledge of all node internals in order to be able to successfully coordinate error recovery for all nodes in a system. That is obviously unwanted.

If we consider all nodes to be composites of a local coordinator and 'other things', then the local coordinator allows for abstraction of things like recovery behaviour (or any behaviour really), we get what you describe (and what many ROS1 nodes now do themselves), while still not muddying a node's life-cycle with coordination concerns.